File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/MenstruationPeriodRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.MenstruationPeriodRecord
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.MenstruationPeriodRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [MenstruationPeriodRecord] to [MenstruationPeriodRecordDto].
 */
internal fun MenstruationPeriodRecord.toDto(): MenstruationPeriodRecordDto =
    MenstruationPeriodRecordDto(
        id = metadata.id,
        metadata = metadata.toDto(),
        startTime = startTime.toEpochMilli(),
        endTime = endTime.toEpochMilli(),
        startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
        endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
    )

/**
 * Converts [MenstruationPeriodRecordDto] to Health Connect [MenstruationPeriodRecord].
 */
internal fun MenstruationPeriodRecordDto.toHealthConnect(): MenstruationPeriodRecord =
    MenstruationPeriodRecord(
        startTime = Instant.ofEpochMilli(startTime),
        endTime = Instant.ofEpochMilli(endTime),
        startZoneOffset = startZoneOffsetSeconds?.let {
            ZoneOffset.ofTotalSeconds(it.toInt())
        } ?: ZoneOffset.UTC,
        endZoneOffset = endZoneOffsetSeconds?.let {
            ZoneOffset.ofTotalSeconds(it.toInt())
        } ?: ZoneOffset.UTC,
        metadata = metadata.toHealthConnect(id),
    )

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/menstruation/menstruation_period_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show MenstruationPeriodRecord, HealthRecordId;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show MenstruationPeriodRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [MenstruationPeriodRecord] to [MenstruationPeriodRecordDto].
@internal
extension MenstruationPeriodRecordToDto on MenstruationPeriodRecord {
  MenstruationPeriodRecordDto toDto() {
    return MenstruationPeriodRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDto(),
    );
  }
}

/// Converts [MenstruationPeriodRecordDto] to [MenstruationPeriodRecord].
@internal
extension MenstruationPeriodRecordDtoToDomain on MenstruationPeriodRecordDto {
  MenstruationPeriodRecord toDomain() {
    return MenstruationPeriodRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      startTime: DateTime.fromMillisecondsSinceEpoch(startTime, isUtc: true),
      endTime: DateTime.fromMillisecondsSinceEpoch(endTime, isUtc: true),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDomain(),
    );
  }
}
